### PR TITLE
[Merged by Bors] - feat: tactic `congr!` and improvement to `convert`

### PR DESCRIPTION
**Pull request by kmill**
This introduces a tactic `congr!` that is an analogue to mathlib 3's `congr'`. It is a more insistent version of `congr` that makes use of more congruence lemmas (including user congruence lemmas), `propext`, `funext`, and `Subsingleton` instances. It also has a feature to lift reflexive relations to equalities. Along with `funext`, the tactic does `intros`, allowing `congr!` to get access to function bodies; the introduced variables can be named using `rename_i` if needed.

This also modifies `convert` to use `congr!` rather than `congr`, which makes it work more like the mathlib3 version of the tactic.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Congr!.lean [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+      let [] ← mvarIdR.apply (← mkConstWithFreshMVarLevels lem) | failure
+      return mvarIdEq
+    if let some mvarId := res then
+      return mvarId
+  return mvarId
+
+/--
+Try to close the goal using `Subsingleton.elim`. Returns whether or not it succeeds.
+-/
+def Lean.MVarId.subsingletonElim (mvarId : MVarId) : MetaM Bool := do
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:103]: Can we do this without assigning universe metavariables? That regression seems very unfortunate.

File: Mathlib/Tactic/Congr!.lean [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        applyCongrThm? mvarId ptype proof
+      if let some mvars := res then
+        return mvars
+    return none
+
+/--
+Try to convert an `Iff` into an `Eq` by applying `iff_of_eq`.
+If successful, returns the new goal, and otherwise returns the original `MVarId`.
+-/
+def Lean.MVarId.iffOfEq (mvarId : MVarId) : MetaM MVarId := do
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:54]: Is this a special case of `liftReflToEq`?

File: Mathlib/Tactic/Congr!.lean [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+-/
+def Lean.MVarId.congrPi? (mvarId : MVarId) : MetaM (Option (List MVarId)) :=
+  observing? do mvarId.apply (← mkConstWithFreshMVarLevels `pi_congr)
+
+/--
+Try to apply `funext`, but only if it is an equality of two functions where at least one is
+a lambda expression.
+One thing this check prevents is accidentally applying `funext` to a set equality, but also when
+doing congruence we don't want to apply `funext` unnecessarily.
+"Obvious" means that the type of the terms being equated is a pi type.
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:126]: Not sure what you mean here? You don't check this, only that one of the terms is a lambda-expression, right?

File: Mathlib/Tactic/Congr!.lean [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean
+import Mathlib.Tactic.Relation.Rfl
+
+/-!
+# The `congr!` tactic
+
+This is a more powerful version of the `congr` tactic.
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:12]: Can you explain (somewhere) how this relates to Lean 3 `congr`/`congr'`/`convert`?

**Review by fpvandoorn — APPROVED** [review]
LGTM

It would be nice to add some tests to the test folder, especially if you think that any of the subfunctions of `congr!` are not yet tested in mathlib.

bors d+

**Comment by fpvandoorn** [comment]
One more remark: `congr! e with ...` might be useful, to name newly introduced variable names. (Doesn't have to go into this PR, of course.)

File: Mathlib/Tactic/Congr!.lean (hunk 3 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean:126]: Oops, that was a remnant of an earlier version. (It was not a very good idea.)

File: Mathlib/Tactic/Congr!.lean (hunk 2 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean:54]: Yeah, it's that but specifically for `iff`, generating a more straightforward proof. Should I do anything more than add a comment about this?

File: Mathlib/Tactic/Congr!.lean (hunk 4 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean]: I added a little bit to relate them.

File: Mathlib/Tactic/Congr!.lean (hunk 1 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean]: I think I found a straightforward way to do it. It synthesizes a `Subsingleton.elim` term using `mkAppM`, which uses `withNewMCtxDepth` so outside metavariables can't be reassigned.

File: Mathlib/Tactic/Congr!.lean (hunk 2 above)
Comment (fpvandoorn) [reply Mathlib/Tactic/Congr!.lean:61]: No, if the generated proof is easier, it is probably nice to keep it separate.

**Comment by kmill** [comment]
There's nothing on the queue right now, so I guess I'll try again and hope timing out was just a fluke?

bors r+

**Pull request closed without merging**